### PR TITLE
add new version of transfer API with B2C support

### PR DESCRIPTION
**Pull request by martinhenningjensen**

### CODE REVIEW

File: domains/transfer/Transfer.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs [hunk 1]
```diff
@@ -70,10 +76,20 @@ protected override void ConfigureWebHost(IWebHostBuilder builder)
...
 
+        builder.UseSetting("B2C:B2CWellKnownUrl",
+            "https://login.microsoftonline.com/d3803538-de83-47f3-bc72-54843a8592f2/v2.0/.well-known/openid-configuration");
+        builder.UseSetting("B2C:ClientCredentialsCustomPolicyWellKnownUrl",
+            "https://datahubeouenerginet.b2clogin.com/datahubeouenerginet.onmicrosoft.com/v2.0/.well-known/openid-configuration?p=B2C_1A_CLIENTCREDENTIALS");
+        builder.UseSetting("B2C:MitIDCustomPolicyWellKnownUrl",
+            "https://datahubeouenerginet.b2clogin.com/datahubeouenerginet.onmicrosoft.com/v2.0/.well-known/openid-configuration?p=B2C_1A_MITID");
+        builder.UseSetting("B2C:Audience", "f00b9b4d-3c59-4c40-b209-2ef87e509f54");
+        builder.UseSetting("B2C:CustomPolicyClientId", "a701d13c-2570-46fa-9aa2-8d81f0d8d60b");
+
```
Comment (tnickelsen) [inline domains/transfer/Transfer.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs:92]: This only flies 'cause we are in a private lib, right?

File: domains/transfer/Transfer.API/API.IntegrationTests/Transfer/Api/Dto/Requests/CreateTransferAgreementProposal20240515ValidatorTest.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+public class CreateTransferAgreementProposal20240515ValidatorTest
+{
+    private CreateTransferAgreementProposal20240515Validator sut = new();
+
+    [Theory]
+    [InlineData("", "ReceiverTin must be 8 digits without any spaces.")]
+    [InlineData("1234567", "ReceiverTin must be 8 digits without any spaces.")]
+    [InlineData("123456789", "ReceiverTin must be 8 digits without any spaces.")]
+    [InlineData("ABCDEFG", "ReceiverTin must be 8 digits without any spaces.")]
+    public void Create_ShouldFail_WhenReceiverTinInvalid(string receiverTin, string expectedContent)
```
Comment (tnickelsen) [inline domains/transfer/Transfer.API/API.IntegrationTests/Transfer/Api/Dto/Requests/CreateTransferAgreementProposal20240515ValidatorTest.cs:17]: Why not hardcode expectedContent?

File: domains/transfer/Transfer.API/API.IntegrationTests/Cvr/Api/Controllers/CvrController20240515Tests.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        var cvrNumbers = new CvrRequestDto(new List<string> { "10150817" });
+        cvrWireMock
+            .Given(Request.Create().WithPath("/cvr-permanent/virksomhed/_search").UsingPost())
+            .RespondWith(Response.Create()
+                .WithStatusCode(200)
+                .WithBodyFromFile("Cvr/Api/Controllers/CvrControllerTests.cvr_response.json")
+            );
+
+        using var client = factory.CreateB2CAuthenticatedClient(Guid.NewGuid(), Guid.NewGuid());
```
Comment (tnickelsen) [inline domains/transfer/Transfer.API/API.IntegrationTests/Cvr/Api/Controllers/CvrController20240515Tests.cs:44]: Why do we need B2C auth on cvr when it's public?

File: domains/transfer/Transfer.API/API.IntegrationTests/Transfer/Api/Controllers/TransferAgreements20240515ControllerTest.cs [hunk 4]
```diff
@@ -0,0 +1,685 @@
...
+using VerifyXunit;
+using Xunit;
+using Xunit.Abstractions;
+using TransferAgreementDto = API.Transfer.Api.Dto.Responses.TransferAgreementDto;
+using TransferAgreementsResponse = API.Transfer.Api.Dto.Responses.TransferAgreementsResponse;
+
+namespace API.IntegrationTests.Transfer.Api.Controllers;
+
+[Collection(IntegrationTestCollection.CollectionName)]
+public class TransferAgreements20240515ControllerTests
```
Comment (tnickelsen) [inline domains/transfer/Transfer.API/API.IntegrationTests/Transfer/Api/Controllers/TransferAgreements20240515ControllerTest.cs:26]: Are these the same tests or any new ones I need to look at?

File: domains/transfer/Transfer.API/API/Transfer/Api/Controllers/TransferAgreementProposal20240515Controller.cs [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+using EnergyOrigin.TokenValidation.b2c;
+using FluentValidation;
+using FluentValidation.AspNetCore;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+
+namespace API.Transfer.Api.Controllers;
+
+[ApiController]
+[Authorize(Policy.B2CCvrClaim)]
```
Comment (tnickelsen) [inline domains/transfer/Transfer.API/API/Transfer/Api/Controllers/TransferAgreementProposal20240515Controller.cs:18]: Why CvrClaim on proposal controller?

File: domains/transfer/Transfer.API/API/Transfer/Api/Controllers/TransferAgreements20240515Controller.cs [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+using EnergyOrigin.TokenValidation.b2c;
+using FluentValidation.AspNetCore;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using ProjectOriginClients;
+
+namespace API.Transfer.Api.Controllers;
+
+[Authorize(Policy.B2CCvrClaim)]
```
Comment (tnickelsen) [inline domains/transfer/Transfer.API/API/Transfer/Api/Controllers/TransferAgreements20240515Controller.cs:21]: Again?

File: domains/transfer/Transfer.API/API.IntegrationTests/Factories/IntegrationTestB2CAuthHandler.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+using System.Threading.Tasks;
+using EnergyOrigin.TokenValidation.b2c;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Net.Http.Headers;
+
+namespace API.IntegrationTests.Factories;
+
+public class IntegrationTestB2CAuthHandler : AuthenticationHandler<AuthenticationSchemeOptions>
```
Comment (tnickelsen) [inline domains/transfer/Transfer.API/API.IntegrationTests/Factories/IntegrationTestB2CAuthHandler.cs:14]: Is this middleware or where is it used?

File: domains/transfer/Transfer.API/API.IntegrationTests/Transfer/Api/Dto/Requests/CreateTransferAgreementProposal20240515ValidatorTest.cs (hunk 2 above)
Comment (martinhenningjensen) [reply domains/transfer/Transfer.API/API.IntegrationTests/Transfer/Api/Dto/Requests/CreateTransferAgreementProposal20240515ValidatorTest.cs]: Moved to assert instead.

**Review by tnickelsen — APPROVED** [review]

**Pull request merged**